### PR TITLE
Fixing Issue #374

### LINKED ISSUE #71: Example file "beam-varying-extrusion-paths" is not working
The file is [here](../blob/development/testdata/IFC4x1ExampleFiles/official/E14_Additional_test_cases/beam-varying-extrusion-paths.ifc). Current version crashes.

### LINKED ISSUE #281: [BUG] Unit test “beam-varying-extrusion-paths” cannot complete subtests PlaneSurfaceViews and VertexViews
Unit test

**Describe the bug**
Correct screenshots are missing in the Unit test “beam-varying-extrusion-paths”

**To Reproduce**
1. Select unit test “beam-varying-extrusion-paths” as ‘Start up Project’
2. Uncomment Code in “beam-varying-extrusion-paths.cpp”
3. Launch debugging 

**Expected behavior**
A unit test should complete all subtests (Including PlaneSurfaceViews and VertexViews) without mistakes. 

**CMake configuration**
- EARLYBINDING_WITH_IFC4x1: selected

**Additional context**
At first #71 Issue should be fixed

### LINKED ISSUE #374: [BUG] beam-revolved-solid.ifc  is not rendered
After several last changes beam-revolved-solid.ifc cannot be rendered anymore. The are several problems regarding this issue:

1.	Placements of points on the circle are not applied.
2.	refId was not positive

**To Reproduce**

1. Launch OpenInfraPlatform
2. Load file beam-revolved-solid.ifc 
3. See error

**Expected behavior**
beam-revolved-solid.ifc should be rendered without any problems.

**Screen shot**
image

**CMake configuration**
- EARLYBINDING_WITH_IFC4x1: selected

### LINKED ISSUE #376: [BUG]  beam-revolved-solid.ifc is rendered incorrectly
**Describe the bug**
Placements of points on the circle are not applied in CurveConverter.h. For this reason beam-revolved-solid.ifc is rendered on the wrong place. 

**To Reproduce**
1. Launch OpenInfraPlatform.
2. Load file 'beam-revolved-solid.ifc'
3. See error

**CMake configuration**
If applicable, list any special configurations you have selected in CMake.
- EARLYBINDING_WITH_IFC4x1: selected

### PR DESCRIPTION
**Pull request by SamuilsRulovs**
Fixes #374 
Fixes #376
Fixes #281
Fixes #71
Applying changes: 

- [ ] Applying placements of points on the circle. 
- [ ] Dealing with warnings "refId was not positive"

### CODE REVIEW

**Review by pjanck — APPROVED** [review]
LFTM.

The second part of the issue should be staged as a separate issue.

**Comment by SamuilsRulovs** [comment]
There are some minor improvements, but, unfortunately, it is still rendered incorrectly  
image

File: Core/src/IfcGeometryConverter/CurveConverter.h [hunk 1]
```diff
@@ -963,11 +963,11 @@ namespace OpenInfraPlatform {
 					
 					// Calculate an angle on the circle (with circle center in (0., 0., 0.)) for trimming begin.
 					carve::geom::vector<3> point = getPointOnCurve<typename IfcEntityTypesT::IfcCircle>(circle, trim1Vec, trimmingPreference);
-					double startAngle = getAngleOnCircle(carve::geom::VECTOR(0., 0., 0.), circleRadius, point);
+					double startAngle = getAngleOnCircle(conicPositionMatrix * carve::geom::VECTOR(0., 0., 0.), circleRadius, point);
 					
 					// Calculate an angle on the circle (with circle center in (0., 0., 0.)) for trimming end.
 					point = getPointOnCurve<typename IfcEntityTypesT::IfcCircle>(circle, trim2Vec, trimmingPreference);
-					double endAngle = getAngleOnCircle(carve::geom::VECTOR(0., 0., 0.), circleRadius, point);
+					double endAngle = getAngleOnCircle(conicPositionMatrix * carve::geom::VECTOR(0., 0., 0.), circleRadius, point);
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/CurveConverter.h]: I'd assume that the function `getAngleOnCircle` does not take into account if the circle is rotated ...

**Review by pjanck — CHANGES_REQUESTED** [review]
suggestion to explore more deeply

File: Core/src/IfcGeometryConverter/CurveConverter.h (hunk 1 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/CurveConverter.h]: Resolved:
0f25f00

File: Core/src/IfcGeometryConverter/CurveConverter.h [hunk 2]
```diff
@@ -957,17 +959,19 @@ namespace OpenInfraPlatform {
 					// **************************************************************************************************************************
 					// Determine position
 					carve::math::Matrix conicPositionMatrix = placementConverter->convertIfcAxis2Placement(circle->Position);
+					carve::math::Matrix inverseConicPositionMatrix;
+					GeomUtils::computeInverse(conicPositionMatrix, inverseConicPositionMatrix);
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/CurveConverter.h]: - change the existing function so that it returns the inverse;
- plus add doxy comments to it;

**Review by pjanck — CHANGES_REQUESTED** [review]

File: Core/src/IfcGeometryConverter/CurveConverter.h (hunk 2 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/CurveConverter.h]: Resolved:
d9ce668

**Review by pjanck — APPROVED** [review]
I suppose that a unit test update will follow in another pull request?

**Comment by SamuilsRulovs** [comment]
Right, I will add unit test update for both ifc here, so this PR could be merged

File: Core/src/IfcGeometryConverter/PlacementConverter.h [hunk 3]
```diff
@@ -346,7 +346,7 @@ namespace OpenInfraPlatform {
                     carve::geom::vector<3>  ref_direction(carve::geom::VECTOR(1.0, 0.0, 0.0)); // defaults to (1.0,0.0) according to the specification
 
                     // interpret Location 
-					translate = convertIfcPoint( axis2placement2d->Location );
+					translate = convertIfcPoint( axis2placement2d->Location.as<typename IfcEntityTypesT::IfcPoint>());
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/PlacementConverter.h]:
```suggestion
					translate = convertIfcPoint( axis2placement2d->Location.template as<typename IfcEntityTypesT::IfcPoint>());
```

Templated functions should be called with the `template` keyword.

File: Core/src/IfcGeometryConverter/SolidModelConverter.h [hunk 4]
```diff
@@ -1155,7 +1155,7 @@ namespace OpenInfraPlatform
 
 					if (axis_placement->Location)
 					{
-						axis_location = placementConverter->convertIfcPoint(axis_placement->Location );
+						axis_location = placementConverter->convertIfcPoint(axis_placement->Location.as<typename IfcEntityTypesT::IfcPoint>());
```
Comment (pjanck) [inline Core/src/IfcGeometryConverter/SolidModelConverter.h]: similar to above

**Review by pjanck — APPROVED** [review]
Looks great!

File: Core/src/IfcGeometryConverter/SolidModelConverter.h (hunk 4 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/SolidModelConverter.h]: Resolved: 09db918

File: Core/src/IfcGeometryConverter/PlacementConverter.h (hunk 3 above)
Comment (SamuilsRulovs) [reply Core/src/IfcGeometryConverter/PlacementConverter.h]: Resolved: 09db918

**Review by pjanck — APPROVED** [review]
Awesome work, I'm really happy seeing 4 issues resolved!

**Pull request merged**